### ISSUE TITLE
Performance issues of goleveldb

### ISSUE THREAD

**Issue by ngaut** [pingcap/tidb#6395]
tidb-server eats lots of CPU inside goleveldb when i am preparing sysbench data 

sysbench /usr/local/Cellar/sysbench/1.0.11/share/sysbench/oltp_read_write.lua --mysql-host=127.0.0.1 --mysql-port=4000 --mysql-user=root --mysql-password=""         --mysql-db=test --tables=2 --table-size=1000000         --report-interval=10         --threads=10 --time=0  prepare

Blow is the profile.
➜  go tool pprof http://localhost:10081/debug/pprof/profile
Fetching profile over HTTP from http://localhost:10081/debug/pprof/profile
Saved profile in /Users/liuqi/pprof/pprof.samples.cpu.004.pb.gz
Type: cpu
Time: Apr 26, 2018 at 3:03pm (CST)
Duration: 30.20s, Total samples = 1.01mins (201.03%)
Entering interactive mode (type "help" for commands, "o" for options)
(pprof) top40 -cum
Showing nodes accounting for 19.74s, 32.52% of 60.70s total
Dropped 528 nodes (cum <= 0.30s)
Showing top 40 nodes out of 234
      flat  flat%   sum%        cum   cum%
         0     0%     0%     24.93s 41.07%  github.com/pingcap/tidb/store/mockstore/mocktikv.(*RPCClient).SendRequest
         0     0%     0%     24.93s 41.07%  github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).SendReq
         0     0%     0%     24.93s 41.07%  github.com/pingcap/tidb/store/tikv.(*RegionRequestSender).sendReqToRegion
         0     0%     0%     24.53s 40.41%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb.(*DB).compactionTransact
         0     0%     0%     23.22s 38.25%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb.(*DB).tCompaction
         0     0%     0%     23.22s 38.25%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb.(*DB).tableAutoCompaction
         0     0%     0%     23.22s 38.25%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb.(*DB).tableCompaction
     0.01s 0.016% 0.016%     23.13s 38.11%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb.(*tableCompactionBuilder).run
         0     0% 0.016%     19.32s 31.83%  internal/poll.(*FD).Write
         0     0% 0.016%     19.32s 31.83%  os.(*File).Write
         0     0% 0.016%     19.32s 31.83%  os.(*File).write
    19.29s 31.78% 31.80%     19.30s 31.80%  syscall.Syscall
         0     0% 31.80%     19.02s 31.33%  syscall.Write
         0     0% 31.80%     19.02s 31.33%  syscall.write
         0     0% 31.80%     17.26s 28.43%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/iterator.(*mergedIterator).Next
         0     0% 31.80%     16.75s 27.59%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/table.(*Writer).writeBlock
         0     0% 31.80%     16.70s 27.51%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb.(*tWriter).append
         0     0% 31.80%     16.69s 27.50%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/table.(*Writer).Append
         0     0% 31.80%     16.66s 27.45%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/table.(*Writer).finishBlock
     0.02s 0.033% 31.83%     15.49s 25.52%  github.com/pingcap/tidb/store/mockstore/mocktikv.newIterator
         0     0% 31.83%     15.40s 25.37%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb.(*tableCompactionBuilder).appendKV
     0.11s  0.18% 32.01%        15s 24.71%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/iterator.(*indexedIterator).setData
         0     0% 32.01%     14.96s 24.65%  github.com/pingcap/tidb/server.(*Server).onConn
         0     0% 32.01%     14.96s 24.65%  github.com/pingcap/tidb/server.(*clientConn).Run
         0     0% 32.01%     14.95s 24.63%  github.com/pingcap/tidb/session.(*session).Execute
         0     0% 32.01%     14.94s 24.61%  github.com/pingcap/tidb/server.(*TiDBContext).Execute
         0     0% 32.01%     14.94s 24.61%  github.com/pingcap/tidb/server.(*clientConn).dispatch
         0     0% 32.01%     14.94s 24.61%  github.com/pingcap/tidb/server.(*clientConn).handleQuery
         0     0% 32.01%     14.36s 23.66%  github.com/pingcap/tidb/store/tikv.(*twoPhaseCommitter).doActionOnBatches.func1
         0     0% 32.01%     14.36s 23.66%  github.com/pingcap/tidb/util/goroutine_pool.(*goroutine).workLoop
     0.08s  0.13% 32.14%     14.32s 23.59%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/table.(*indexIter).Get
     0.05s 0.082% 32.22%     14.13s 23.28%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/table.(*Reader).getDataIterErr
     0.05s 0.082% 32.31%     13.87s 22.85%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/table.(*Reader).getDataIter
         0     0% 32.31%     13.82s 22.77%  github.com/pingcap/tidb/session.(*session).executeStatement
         0     0% 32.31%     13.82s 22.77%  github.com/pingcap/tidb/session.runStmt
     0.09s  0.15% 32.45%     11.45s 18.86%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/table.(*Reader).readBlockCached
     0.04s 0.066% 32.52%     10.76s 17.73%  github.com/pingcap/tidb/vendor/github.com/pingcap/goleveldb/leveldb/iterator.(*indexedIterator).Next
         0     0% 32.52%     10.67s 17.58%  github.com/pingcap/tidb/session.(*TxnState).Commit
         0     0% 32.52%     10.67s 17.58%  github.com/pingcap/tidb/session.(*session).CommitTxn
         0     0% 32.52%     10.67s 17.58%  github.com/pingcap/tidb/session.(*session).doCommit

**Comment by ngaut** [comment]
[pprof.samples.cpu.004.pb.gz](https://github.com/pingcap/tidb/files/1950450/pprof.samples.cpu.004.pb.gz)

**Comment by winoros** [comment]
Seems that testing with mock-tikv instead of tikv?

**Comment by ngaut** [comment]
Yes, I used mock-tikv engine, this is not a high priority issue. Take your time.

**Comment by winkyao** [comment]
This is a known issue, @ngaut, goleveldb use skiplist to implemnt the kv storage. It uses two buffer mainly, one call kvdata, which is used to store key-value, and other is used to store meta data. the kv data array is pre-allocated, but metadata is not. Also, I found the FindGE costs too many cpu in the frame graph.

Maybe we need to find a replacement, or we implement a new memdb?

**Comment by ngaut** [comment]
There is a better implementation, see https://github.com/coocood/badger/blob/master/skl/skl.go#L127
